partitura/musicanalysis/meter.py

```diff
@@ -0,0 +1,310 @@
+        for c in self.clusters:
+            i = 0
+            while i < len(onsets) and onsets[i] < INIT_DURATION:
+                a = Agent()
+                a.beat_interval = c.interval
+                a.history.append((onsets[i], salience[i]))
+                a.prediction = onsets[i] + c.interval
+                a.score = salience[i]
+                self.agents.append(a)
+                i += 1
+
+    def track(self, onsets, salience):  
+        for e_i in range(len(onsets)):
+            e = onsets[e_i]
+            new_agents = []
+            remove_agents = []
+            for a in self.agents:
+                if e - a.lastBeat() > TIMEOUT:
+                    remove_agents.append(a)
+                else:
+                    while a.prediction + TOLERANCE_POST*a.beat_interval < e:
+                        a.history.append((a.prediction, 0)) 
+                        a.prediction += a.beat_interval
+                    if a.prediction - TOLERANCE_PRE*a.beat_interval <= e and e <= a.prediction + TOLERANCE_POST*a.beat_interval:
+                        if abs(a.prediction - e) > TOLERANCE_INNER:
+                            a_new = Agent()
+                            a_new.beat_interval = a.beat_interval
+                            a_new.history = a.history[:]
+                            a_new.prediction = a.prediction
+                            a_new.score = a.score
+                            new_agents.append(a_new)
+                        err = e - a.prediction
+                        a.beat_interval = a.beat_interval + err*CORRECTION_FACTOR
+                        a.prediction = e + a.beat_interval
+                        a.history.append((e, salience[e_i]))
+                        a.score += (1-abs(err/a.beat_interval)/2.) * salience[e_i]
+
+            for a in remove_agents:
+                self.agents.remove(a)
+            self.agents = self.agents + new_agents
+
+            # remove duplicate agents
+            duplicate = np.zeros(len(self.agents))
+            agents_all = self.agents[:]
+            self.agents = []
+            for i in range(len(agents_all)):
+                for j in range(i+1, len(agents_all)):
+                    if duplicate[i] > 0 or duplicate[j] > 0:
+                        continue
+
+                    if abs(agents_all[i].beat_interval - agents_all[j].beat_interval) < 0.01 \
+                            and abs(agents_all[i].lastBeat() - agents_all[j].lastBeat()) < 0.02:
+                        if agents_all[i].score > agents_all[j].score:
+                            duplicate[j] += 1
+                        else:
+                            duplicate[i] += 1
+                            break
+
+            self.agents = sorted(np.asarray(agents_all)[(duplicate < 1)].tolist(
+            ), key=lambda x: x.score, reverse=True)[:MAX_AGENTS]
+
+        self.agents = sorted(self.agents, key=lambda x: x.score, reverse=True)
+
+
+class Cluster():
```

Comment:
Class Cluster may need some documentation. Also a line describing the ioi argument.

Reply:
I add documentation.